core/Pos/Util/Concurrent/PriorityLock.hs

```diff
@@ -65,17 +67,25 @@ unlockP (PriorityLock vstate) =
     atomically $ readTVar vstate >>= \case
         Unlocked -> error "Pos.Util.PriorityLock.unlockP: lock is already unlocked"
 
-        Locked [] [] ->
-            -- no one is waiting on the lock, so it will be unlocked now
-            writeTVar vstate Unlocked
+        Locked hwaiters lwaiters
+            | empty hwaiters && empty lwaiters ->
+              -- no one is waiting on the lock, so it will be unlocked now
+              writeTVar vstate Unlocked
 
-        Locked (waiter:hwaiters) lwaiters -> do
-            writeTVar vstate (Locked hwaiters lwaiters)
-            putTMVar waiter ()
+        Locked hwaiters lwaiters
+            | empty hwaiters -> do
+                  -- no one is waiting with high priority, so we
+                  -- dequeue from the low priority waiters
+                  let (waiter, lwaiters') = dequeue' lwaiters
+                  writeTVar vstate (Locked hwaiters lwaiters')
+                  putTMVar waiter ()
 
-        Locked [] (waiter:lwaiters) -> do
-            writeTVar vstate (Locked [] lwaiters)
+        Locked hwaiters lwaiters -> do
+            -- dequeue from the high priority waiters
+            let (waiter, hwaiters') = dequeue' hwaiters
+            writeTVar vstate (Locked hwaiters' lwaiters)
             putTMVar waiter ()
+    where dequeue' q = fromMaybe (error "Logic error in PriorityLock") (dequeue q)
```

Comment:
I think we could restructure this to avoid this error case by avoiding the `empty` operation.

This is exactly analogous to the general advice to pattern match on [] & (:) rather than use empty, head and tail.

Something like:
```
Locked hwaiters lwaiters
    -- We have at least one waiter with high priority
  | Just (waiter, hwaiters') <- dequeue hwaiters
 ->  ...
    -- No one is waiting with high priority, but we
    -- do have a low priority waiter
  | Just (waiter, lwaiters') <- dequeue lwaiters
 -> ...

    -- No high or low priority waiters. No one is waiting
    -- on the lock, so it will be unlocked now
  | otherwise
 -> ...
```

Reply:
Thanks, that's much better!